src/storage/garbage_collector.cpp

```diff
@@ -129,54 +243,279 @@ void GarbageCollector::ProcessDeferredActions() {
-  while (true) {
+  while (next != nullptr && active_txns_iter != active_txns->end()) {
+    if (transaction::TransactionUtil::NewerThan(*active_txns_iter, curr->Timestamp().load())) {
+      // curr is the version that *active_txns_iter would be reading
+      active_txns_iter++;
+    } else if (transaction::TransactionUtil::NewerThan(next->Timestamp().load(), *active_txns_iter)) {
+      // curr is visible to txns traversed till now, so can't GC curr. next is not visible to any txn, attempt GC
+      // Since *active_txns_iter is not reading next, that means no one is reading this
+      // And so we can reclaim next
+      if (interval_length == 0) {
+        // This only happens when the first compaction is initiated. Other compactions begin in Case 3.
+        // This is the first undo record to compact
+        BeginCompaction(&start_record, curr, next, &interval_length);
+      } else {
+        // Process the undo record to determine the set of columns for the compacted record.
+        ReadUndoRecord(start_record, next, &interval_length);
+      }
+      // Update curr
+      curr = curr->Next();
+    } else {
+      if (interval_length > 1) {
+        // Create the undo record by a second traversal through the records to be compacted.
+        UndoRecord *compacted_undo_record = CreateUndoRecord(start_record, next);
+        // Copy the varlen attributes of the compacted record
+        CopyVarlen(compacted_undo_record);
+        // Link the compacted record to the version chain.
+        LinkCompactedUndoRecord(start_record, &curr, next, compacted_undo_record);
+      }
+      // Set interval length to 0
+      EndCompaction(&interval_length);
+      // Begin compaction for the next interval
+      BeginCompaction(&start_record, curr, next, &interval_length);
+      // curr was not claimed in the previous iteration, so possibly someone might use it
+      curr = curr->Next();
+    }
     next = curr->Next();
-    // This is a legitimate case where we truncated the version chain but had to restart because the previous head
-    // was aborted.
-    if (next == nullptr) return;
-    if (transaction::TransactionUtil::NewerThan(oldest, next->Timestamp().load())) break;
-    curr = next;
-  }
-  // The rest of the version chain must also be invisible to any running transactions since our version
-  // is newest-to-oldest sorted.
+  }
+
+  if (interval_length > 1) {
+    // Create the undo record by a second traversal through the records to be compacted.
+    UndoRecord *compacted_undo_record = CreateUndoRecord(start_record, next);
+    // Copy the varlen attributes of the compacted record
+    CopyVarlen(compacted_undo_record);
+    // Link the compacted record to the version chain.
+    LinkCompactedUndoRecord(start_record, &curr, next, compacted_undo_record);
+  } else if (interval_length == 1) {
+    if (transaction::TransactionUtil::NewerThan(active_txns->back(), curr->Timestamp().load())) {
```

Comment:
`active_txns->back()` has undefined behaviour if `active_txns` is empty, you might want to check for that.

Reply:
Transaction Manager ensures that if the active_txns is not empty, it inserts a timestamp.